client/history/peer_resolver.go

```diff
@@ -110,3 +115,54 @@ func (pr peerResolver) GetAllPeers() ([]string, error) {
 	}
 	return peers, nil
 }
+
+func (pr peerResolver) GlobalRatelimitPeers(ratelimits []string) (map[string][]string, error) {
+	// History was selected simply because it already has a ring and an internal-only API.
+	// Any service should be fine, it just needs to be shared by both ends of the system.
+	hosts, err := pr.resolver.Members(service.History)
+	if err != nil {
+		return nil, fmt.Errorf("unable to get history peers: %w", err)
+	}
+	if len(hosts) == 0 {
+		// can occur when shutting down the only instance because it calls EvictSelf ASAP.
+		// this is common in dev, but otherwise *probably* should not be possible in a healthy system.
+		return nil, fmt.Errorf("unable to get history peers: no peers available")
+	}
+
+	results := make(map[string][]string, len(hosts))
+	initialCapacity := len(ratelimits) / len(hosts)
+	// add a small buffer to reduce copies, as this is only an estimate
+	initialCapacity += initialCapacity / 10
+	// but don't use zero, that'd be pointless
+	if initialCapacity < 1 {
+		initialCapacity = 1
+	}
+
```

Comment:
+1, I'd not bother maybe as part of this PR, but at allowing hostnames to be mapped to a particular rate-limit (via sampling or debug logs) would probably be operationally be quite helpful in diagnosing any problems

Reply:
since I broadly agree but I think there are a few overlapping uses:
what kind of goals do you think are worth building immediately vs "soon"?

OTOH I think there are:
- if we think there's a sharding error logic of some kind, we'd need to know keys and mapped peers.  sampling might _luckily_ log the issue, but this isn't hit all that frequently so like 1% might take several minutes until host X tells us what we need... and I'm not really sure that's good enough tbh.  but it feels like too much info to log all the time.
- if we want to sanity check the behavior, sampling works fine.  we should see evidence in other places too, e.g. wrong allow/reject rates, but sampled logs let us ignore it for a bit and check later and that's kinda convenient.
- ^ both of these are possibly solved by a debug page that dumps collection info / usage / etc, for _single host_ investigating (but not bulk), and it'd be nice to have a system to allow that in open-source.  I don't believe we have any right now though.

and... feels like maybe other scenarios are useful?  I haven't been involved in much of our ring issues, not sure what worked / would have helped in those.